src/libcore/iter.rs

```diff
@@ -115,7 +119,10 @@ pub trait Iterator {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     fn count(self) -> usize where Self: Sized {
-        self.fold(0, |cnt, _x| cnt + 1)
+        self.fold(0, |cnt, _| match cnt.checked_add(1) {
+            Some(c) => c,
+            None => panic!("overflow while counting the elements of an iterator"),
```

Comment:
I think @tbu-'s change here is fine.

I'd be willing to take the speed hit here.

(The one thing I *might* consider trying to change, and this would be a big thing, would be to change `count`'s signature to `u64` rather than `usize`.  I'm basically thinking about iterators that are traversing ranges rather than collections of word-aligned objects in memory.)